### ISSUE TITLE
Database disk image is malformed

### ISSUE THREAD

**Issue by devos50** [Tribler/tribler#5252]
From the error reporter (7.5.0-RC2):

```python
Traceback (most recent call last):
File "src\tribler-gui\tribler_gui\core_manager.py", line 106, in on_received_state
RuntimeError: database disk image is malformed
--LONG TEXT--
Traceback (most recent call last):
File "src\run_tribler.py", line 95, in start_tribler
File "src\tribler-core\tribler_core\session.py", line 440, in start
File "src\tribler-core\tribler_core\session.py", line 170, in load_ipv8_overlays
File "ipv8\attestation\trustchain\community.py", line 64, in __init__
File "ipv8\attestation\trustchain\database.py", line 43, in __init__
File "ipv8\attestation\trustchain\database.py", line 419, in open
File "ipv8\database.py", line 127, in open
File "ipv8\database.py", line 162, in _initial_statements
sqlite3.DatabaseError: database disk image is malformed
--CONTEXT--
{'message': 'Task exception was never retrieved', 'exception': DatabaseError('database disk image is malformed'), 'future': <Task finished coro=<start_tribler_core.<locals>.start_tribler() done, defined at src\run_tribler.py:64> exception=DatabaseError('database disk image is malformed')>}
```

Related to #1993

**Comment by qstokkink** [comment]
What should we do when the database is corrupt beyond repair? Remove it?

**Comment by ichorid** [comment]
In this case, we should mark the database corrupted and schedule a DB check on the next startup. Ideally, we should employ DB's own mechanisms for that.

**Comment by devos50** [comment]
@ichorid do we want to fix this issue in 7.5?

**Comment by ichorid** [comment]
There were only 2 bugreports with this kind of bug in the last few months. One should put some _real_ effort to garble their SQLite database to produce this kind of error. It is alson not trivial to recover from it later. I would mark it a low-priority and moved to backlog.

**Comment by devos50** [comment]
On our (Proxmoxed) Mac, I now have a corrupt `metadata.db`, which is crashing consistently a few seconds after start-up. It errors with different kind of errors actually, depending on what the interactions with the database are. For example:

```
Traceback (most recent call last):
  File "ipv8/taskmanager.py", line 131, in done_cb
  File "tribler_core/modules/metadata_store/community/remote_query_community.py", line 149, in on_remote_select_response
  File "tribler_core/modules/metadata_store/store.py", line 287, in process_compressed_mdblob_threaded
  File "concurrent/futures/thread.py", line 57, in run
  File "tribler_core/modules/metadata_store/store.py", line 276, in _process_blob
  File "tribler_core/modules/metadata_store/store.py", line 295, in process_compressed_mdblob
  File "tribler_core/modules/metadata_store/store.py", line 328, in process_squashed_mdblob
  File "<string>", line 2, in process_payload
  File "pony/orm/core.py", line 515, in new_func
  File "tribler_core/modules/metadata_store/store.py", line 409, in process_payload
  File "pony/orm/core.py", line 3974, in get
  File "pony/orm/core.py", line 4077, in _find_one_
  File "pony/orm/core.py", line 4136, in _find_in_db_
  File "pony/orm/core.py", line 938, in _exec_sql
  File "pony/orm/core.py", line 1805, in prepare_connection_for_query_execution
  File "pony/orm/core.py", line 1891, in flush
  File "pony/orm/core.py", line 5393, in _save_
  File "pony/orm/core.py", line 5240, in _save_created_
  File "pony/utils/utils.py", line 106, in throw
pony.orm.core.UnexpectedError: Object ChannelMetadata[new:70] cannot be stored in the database. DatabaseError: database disk image is malformed
```

I will save this database image.

**Comment by xoriole** [comment]
Adding one more stack trace from the error reporter for this issue
```python
Traceback (most recent call last):
File "src\tribler-gui\tribler_gui\core_manager.py", line 68, in on_core_finished
RuntimeError: The Tribler core has unexpectedly finished with exit code -1073741819 and status: 1!

[PID:7780] 2020-10-22 00:24:14,067 - ERROR <taskmanager:135> RemoteQueryCommunity.done_cb(): Task resulted in error: database disk image is malformed
Traceback (most recent call last):
File "lib\site-packages\pony\orm\dbapiprovider.py", line 55, in wrap_dbapi_exceptions
File "lib\site-packages\pony\orm\dbapiprovider.py", line 279, in execute
sqlite3.DatabaseError: database disk image is malformed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "ipv8\taskmanager.py", line 131, in done_cb
File "src\tribler-core\tribler_core\modules\metadata_store\community\remote_query_community.py", line 149, in on_remote_select_response
File "src\tribler-core\tribler_core\modules\metadata_store\store.py", line 287, in process_compressed_mdblob_threaded
File "lib\concurrent\futures\thread.py", line 57, in run
File "src\tribler-core\tribler_core\modules\metadata_store\store.py", line 278, in _process_blob
File "lib\site-packages\pony\orm\core.py", line 476, in __exit__
File "lib\site-packages\pony\orm\core.py", line 490, in _commit_or_rollback
File "lib\site-packages\pony\orm\core.py", line 381, in commit
File "lib\site-packages\pony\orm\core.py", line 370, in rollback_and_reraise
File "lib\site-packages\pony\utils\utils.py", line 95, in reraise
File "lib\site-packages\pony\orm\core.py", line 379, in commit
File "lib\site-packages\pony\orm\core.py", line 1896, in flush
File "lib\site-packages\pony\orm\core.py", line 5421, in _save_
File "lib\site-packages\pony\orm\core.py", line 5330, in _save_updated_
File "lib\site-packages\pony\orm\core.py", line 949, in _exec_sql
File "lib\site-packages\pony\orm\core.py", line 1784, in reconnect
File "lib\site-packages\pony\utils\utils.py", line 95, in reraise
File "lib\site-packages\pony\orm\core.py", line 947, in _exec_sql
File "<string>", line 2, in execute
File "lib\site-packages\pony\orm\dbapiprovider.py", line 74, in wrap_dbapi_exceptions
pony.orm.dbapiprovider.DatabaseError: database disk image is malformed

(Timestamp: 1603293855257, traceback timestamp: 1603293854068)
```
[7.5.2](http://reporter.tribler.org/stacktrace/36039) [7.5.3](http://reporter.tribler.org/stacktrace/36050)

**Comment by ichorid** [comment]
Most probably, DB corruption is a result of interrupted/crashed DB upgrade/conversion procedure. Safety measures are disabled doing DB upgrade for speed.

**Comment by xoriole** [comment]
Looking at the uptime and the logs 
[7.5.2](http://reporter.tribler.org/stacktrace/36039) - 1318.8 seconds
[7.5.3](http://reporter.tribler.org/stacktrace/36050) - 179.57 seconds

I'm not sure if it was during upgrade

**Comment by ichorid** [comment]
I'll try fixing it for 7.6. @devos50 , could you please send me the corrupted DB if it is not that big?

**Comment by devos50** [comment]
@ichorid it seems that I saved that database on our server named "proxmox" and on the VM with ID 105. You can find the corrupt metadata database in the following path: `/Users/tribler/tribler-failing-metadata/db/7.5/sqlite/metadata.db`.

**Comment by ichorid** [comment]
>Probably we need to provide an option to re-create the database from scratch for these cases

thanks for suggestion, @sentry-for-tribler bot! :wink: 

Jokes aside, I'm going to add a subsystem to schedule a DB check after this kind of crashes.

**Comment by rgoshko** [comment]
Got the same issue today when my Mac mini unexpectedly rebooted:

`Traceback (most recent call last):
  File "tribler_gui/core_manager.py", line 123, in on_received_state
RuntimeError: database disk image is malformed
--LONG TEXT--
Traceback (most recent call last):
  File "pony/orm/dbapiprovider.py", line 53, in wrap_dbapi_exceptions
  File "pony/orm/dbapiprovider.py", line 269, in execute
sqlite3.DatabaseError: database disk image is malformed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "run_tribler.py", line 81, in start_tribler
  File "tribler_core/session.py", line 425, in start
  File "tribler_core/modules/metadata_store/store.py", line 135, in __init__
  File "pony/orm/core.py", line 1138, in generate_mapping
  File "<string>", line 2, in check_tables
  File "pony/orm/core.py", line 528, in new_func
  File "pony/orm/core.py", line 1211, in check_tables
  File "pony/orm/dbschema.py", line 78, in check_tables
  File "<string>", line 2, in execute
  File "pony/orm/dbapiprovider.py", line 69, in wrap_dbapi_exceptions
pony.orm.dbapiprovider.DatabaseError: database disk image is malformed

--CONTEXT--
{'message': 'Task exception was never retrieved', 'exception': DatabaseError('database disk image is malformed'), 'future': <Task finished name='Task-1' coro=<start_tribler_core.<locals>.start_tribler() done, defined at run_tribler.py:50> exception=DatabaseError('database disk image is malformed')>}`

I am unsure which db file is "bad", in ~/.Tribler/7.5/sqlite I have the following two db files from today:

`    32829440  6 Nov 11:28 metadata.db
   404717568  6 Nov 11:27 trustchain.db`

Thanks.

**Comment by ichorid** [comment]
@rgoshko , thanks for reporting! It is caused by `metadata.db`. Only that one is handled by Pony ORM in 7.5.x series.

You can try to recover it by creating a dump. Or just delete it (but you will lose your personal channel).

We will add a procedure to handle this kind of failures in 7.6.

**Comment by rgoshko** [comment]
Thanks @ichorid , I removed the `metadata.db` file and I am back up and running now.

**Comment by ichorid** [comment]
related to #1993

**Comment by kozlovsky** [comment]
Fixed in #7628
